### PR TITLE
Tm 643 contract detail hours

### LINKED ISSUE #643: Contract detail should show a column of total hours worked before contract start date
For example, a contract has 2 projects, with 150 hours worked thus far. The first set of contract hours, shows 125 approved hours. A new contract is pending that begins on Jan 1 and the current date is Jan 5. The new column should display the amount of hours worked up until the Jan 1 date.

To clarify, the total should be the running total for all entries on the associated projects up until the start date of the contract hour for each contract hour.

### PR DESCRIPTION
**Pull request by robline**
Fixes #643

### CODE REVIEW

File: timepiece/contracts/models.py [hunk 1]
```diff
@@ -60,6 +60,30 @@ def get_admin_url(self):
     def get_absolute_url(self):
         return reverse('view_contract', args=[self.pk])
 
+    def get_noncontract_entries(self, entries=None):
```
Comment (daaray) [inline timepiece/contracts/models.py]: having a default value of `None` will cause this method to to throw an `AtrtributeError` if no entries are passed in:

```
----> 1 pc.get_noncontract_entries()

...models.py in get_noncontract_entries(self, entries)
     71
     72         for contract in contracts:
---> 73             entries.exclude(
     74                 project__in=contract.projects.all(),
     75                 start_time__gte=contract.start_date,

AttributeError: 'NoneType' object has no attribute 'exclude'

```

File: timepiece/contracts/models.py [hunk 2]
```diff
@@ -60,6 +60,30 @@ def get_admin_url(self):
...
+            entries.exclude(
+                project__in=contract.projects.all(),
+                start_time__gte=contract.start_date,
+                end_time__lt=contract.end_date + relativedelta(days=1)
+                )
+        return entries
+
+    @property
+    def pre_launch_entries(self):
+        pce = Entry.objects.filter(
```
Comment (daaray) [inline timepiece/contracts/models.py]: very ticky tacky, but can we rename `pce` to `entries` where it is used throughout this PR.  Or `projectcontract_entries`, if you want to be verbose.

File: timepiece/contracts/tests/test_contracts.py [hunk 3]
```diff
@@ -1,11 +1,14 @@
 import datetime
+from dateutil.relativedelta import relativedelta
```
Comment (daaray) [inline timepiece/contracts/tests/test_contracts.py]: This should be revised a bit; whole lib imports should be first, then partial imports:

```
import datetime
import mock

from dateutil.relativedelta import relativedelta

...
```

File: timepiece/templates/timepiece/contract/view.html [hunk 4]
```diff
@@ -56,6 +64,75 @@
...
+                        <td>
+                            {{ contract.pre_launch_hours_worked|floatformat:2 }}
+                            ({% widthratio contract.pre_launch_hours_worked contract.contracted_hours 100 %}%)
+                            <a class="btn btn-danger btn-mini" data-toggle="popover"
+                               data-html="true"
+                               title="Precontract Entries"
+                               data-content="
+                                   <table class='table table-bordered'>
+                                      <tr>
+                                        <td class='table'>Project</td>
```
Comment (daaray) [inline timepiece/templates/timepiece/contract/view.html]: As we discussed in person, I think having the Entry's activity and hours here would be beneficial to help the PM/Viewer determine how to handle these outliers

**Comment by robline** [comment]
PR comments addressed. Thanks

**Comment by daaray** [comment]
:chicken:

**Pull request merged**